### DISCUSSION TITLE
metadata csv esport

### DISCUSSION THREAD

**Discussion by jetmcquack** [RhetTbull/osxphotos#1407] [Q&A]
Hello, I'm trying to make a big upload on an Omeka Server and to do that I need a single big csv file which contains each photo as a single row and the exif datas as columns.... I can't find out if there's a way to that with osxphotos....

**Comment by RhetTbull** [comment]
Yes. See the help for [osxphotos query](https://ofek.dev/pyapp/latest/) or `osxphotos help query`. In particular, you want the `--field` option.  The basic format is `--field FIELD TEMPLATE` where FIELD is the name you want in the CSV file and TEMPLATE is an [osxphotos template string](https://rhettbull.github.io/osxphotos/template_help.html) describing what data to put in that column. 

If you want access to all the EXIF data in the image files beyond what's included in the Photos database, you'll need to use the `{exiftool}` template and have [exiftool](https://exiftool.org/) installed.

Here's an example. I don't know what info you really need so you'll have to play with this. This prints the title and description from Photos then also grabs the `EXIF:Make` field using `exiftool`:

`osxphotos query --field name "{original_name}" --field title "{title}" --field description "{descr}" --field make "{exiftool:EXIF:Make}"`

**Reply by RhetTbull** [reply]
The above command produces a CSV like this:

```
name,title,description,make
Pumpkins3,_,_,Canon
IMG_3092,_,_,Apple
Tulips,Tulips tied together at a flower shop,Wedding tulips,FUJIFILM
IMG_4547,Elder Park,"Elder Park, Adelaide, Australia",Apple
```

The `_` character is used for a missing value but you can change this using the default value of the template system (see the [template help](https://rhettbull.github.io/osxphotos/template_help.html)).  For example, if you wanted an empty column instead:

`osxphotos query --field name "{original_name,}" --field title "{title,}" --field description "{descr,}" --field make "{exiftool:EXIF:Make,}"`

**Comment by jetmcquack** [comment]
Thanks this is really helpful, but as far as I understand this generate a csv file querying the Photos db, not the export folder with the exported files generated by osxphotos , correct? 
I’m asking this because I need to generate the file according to the file name expected in the export function, and not the original one inside photos… if this is not possible I’ll import the photos again in photos in a new db using the new filenames, in order to generate the correct csv

**Reply by RhetTbull** [reply]
Because everything in the `osxphotos query  --field` is a template, you can use the same template you used for export to generate a CSV that matches the export. What did you use for `--directory` and `--filename` for your export? Let's assume you used `--directory {folder_album,NoAlbum}` and `--filename {original_name}`, then in the query you'd do:

`osxphotos query --field name "{folder_album,NoAlbum}/{original_name}" --field title "{title,}" --field description "{descr,}" --field make "{exiftool:EXIF:Make,}"`

If you used a filter on the export (e..g `--added-after DATE`), you'd add the same filter to the query command.

If you are proficient in python you could write your own script that reads the export database created by osxphotos to generate this after the fact.  This is left as an exercise for the reader but I'll give a hint: `export_db.py`: `get_uuid_for_file()` and `get_photoinfo_for_uuid()`

**Comment by ltcarter47** [comment]
Hi, I'm trying to do something similar here, I want a CSV file with some data on my photos so I do a bit of analysis on my library.  I have been successful in getting an output, but I am having trouble getting the filesize to output with it.  I started with your examples above and have been playing with the last field to see if I can get the photos size.  With the following I get a "_" for filesize on every image.
`osxphotos query --field name "{original_name}" --field title "{title}" --field description "{descr}" --field size "{exiftool:imagesize}"`

I've also tried this but it gave me an error:
`osxphotos query --field name "{original_name}" --field title "{title}" --field description "{descr}" --field size "{original_filesize}"`

> ValueError: Unmatched template variables in field size {original_filesize}

Is there a way to use query and get the file size in bytes as one of the "columns"?

On an adjacent topic, is there a way to get this to actually write a csv file?  The above outputs straight into the terminal window and needs to be manually saved with some minor cleanup (getting rid of the terminal stuff before and after the actual output).

Thanks for any assistance you can offer!

**Reply by RhetTbull** [reply]
Hi @ltcarter47  You can do this with `{photo.original_filesize}`.  For frequently used template fields, OSXPhotos provides template fields pre-defined such as `{keywords}`.  For anything not already defined, you can access the photo object directly using `{photo.field_name`}. So for your example, the following would work: 

`osxphotos query --field name "{original_name}" --field title "{title}" --field description "{descr}" --field size "{photo.original_filesize}"`

> On an adjacent topic, is there a way to get this to actually write a csv file? 

Yes, use output redirection:

`osxphotos query --field name "{original_name}" --field title "{title}" --field description "{descr}" --field size "{photo.original_filesize}" > results.csv`

The `> filename` sends the output of the command to the given file. This is a standard part of the Unix shell used by macOS, not something that OSXPhotos does.

To do this with the `{exiftool}` template, you need to use the full `Group:Tag` naming convention. For example, this would work:

`osxphotos query --field name "{original_name}" --field title "{title}" --field description "{descr}" --field size "{exiftool:File:FileSize}"`

**Reply by ltcarter47** [reply]
Wonderful, worked like a charm!  Thanks for your quick response and for making such a useful tool, this is going to help me out a lot!
